### PR TITLE
Backwards Compatible Generic Modules

### PR DESCRIPTION
**Pull request by SentryMan**
Adds all the backwards compatibility stuff from #566 minus the SPI changes

### CODE REVIEW

File: inject/src/test/java/io/avaje/inject/LegacyBeanScopeBuilderTest.java [hunk 1]
```diff
@@ -0,0 +1,338 @@
...
+  private TDModule bc(String name, Class<?>[] provides, Class<?>[] requires) {
+    return bc(name, provides, requires, new Class[0]);
+  }
+
+  private TDModule bc(
+      String name, Class<?>[] provides, Class<?>[] requires, Class<?>[] requiresPkg) {
+    return new TDModule(name, provides, requires, requiresPkg);
+  }
+
+  private static class TDModule implements AvajeModule {
```
Comment (rbygrave) [inline inject/src/test/java/io/avaje/inject/LegacyBeanScopeBuilderTest.java:201]: My guess is that this should implement Module and not AvajeModule

**Review by rbygrave — APPROVED** [review]

**Review by rbygrave — CHANGES_REQUESTED** [review]
Need to resolve why InjectPlugin exists

**Comment by rbygrave** [comment]
Did you hit this issue of desiring 10.x to be backwards compatible or did someone else?  Just wondering of the pain level. Also just to say I'm kind of not loving "AvajeModule" as a name ... so I'm hoping that something better might come to me.

**Comment by SentryMan** [comment]
Yeah, I noticed that all the plugins and libraries I've made at work ceased to function

**Comment by SentryMan** [comment]
> Need to resolve why InjectPlugin exists

The maven plugin/generator is unable to get the type array from legacy plugins.

**Comment by SentryMan** [comment]
> don't like "AvajeModule" as a name 

I was thinking of `AvajeModule` sort of like Guice has `GuiceModule`.

I got a couple of alternate options (just spitballing here):

- BeanModule
- ScopeModule
- GraphModule
- WiringModule
- DependencyModule

**Comment by rob-bygrave** [comment]
> AvajeModule sort of like Guice has GuiceModule

Yeah, that makes sense. IMO the original "Module" was a poor choice by me. "InjectModule" is taken by the annotation (which seems ok) ... so I'm definately warming to AvajeModule (it's mostly used internally).

I should get to review this again today. Great work by the way !!!

**Pull request merged**